lib/ipip.c

```diff
@@ -0,0 +1,141 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <rte_ip.h>
+#include <rte_mbuf.h>
+#include <rte_memcpy.h>
+#include <rte_byteorder.h>
+
+#include "gatekeeper_ipip.h"
+
+#define IP_VERSION              (0x40)
+/* Default IP header length == five 32-bits words. */
+#define IP_HDRLEN               (0x05)
+/* From RFC 1340. */
+#define IP_DEFTTL               (64)
+#define IP_VHL_DEF              (IP_VERSION | IP_HDRLEN)
+#define IP_DN_FRAGMENT_FLAG     (0x0040)
+
+#define IPv6_DEFAULT_VTC_FLOW   (0x60000000)
+#define IPv6_DEFAULT_HOP_LIMITS (0xFF)
+
+int
+encapsulation(struct rte_mbuf *pkt, uint8_t priority,
+	struct ipip_tunnel_info *info)
+{
+	uint64_t ol_flags = 0;
+
+	struct ether_hdr *inner_eth;
+	struct ether_hdr *outer_eth;
+	struct ipv4_hdr *outer_ip4hdr;
+	struct ipv6_hdr *outer_ip6hdr;
+
+	inner_eth = rte_pktmbuf_mtod(pkt, struct ether_hdr *);
+
+	if (info->flow.proto == ETHER_TYPE_IPv4) {
+		/* Allocate space for outer IPv4 header. */
+		outer_eth = (struct ether_hdr *) rte_pktmbuf_prepend(pkt,
+				sizeof(struct ipv4_hdr));
+		if (!outer_eth) {
+			RTE_LOG(ERR, MBUF,
+				"Not enough headroom space in the first segment!\n");
+			return -1;
+		}
+
+		outer_ip4hdr = (struct ipv4_hdr *) &outer_eth[1];
+
+		/* Fill up the outer Ethernet header. */
+		rte_memcpy(&outer_eth->s_addr, &inner_eth->d_addr,
+				sizeof(struct ether_addr));
+		/* Fill up the destination MAC address via Gateway MAC. */
+		rte_memcpy(&outer_eth->d_addr, &info->gateway_mac,
+				sizeof(struct ether_addr));
+
+		outer_eth->ether_type = rte_cpu_to_be_16(ETHER_TYPE_IPv4);
+
+		/* Fill up the outer IP header. */
+		outer_ip4hdr->version_ihl = IP_VHL_DEF;
+		outer_ip4hdr->type_of_service = (priority << 2);
+		outer_ip4hdr->packet_id = 0;
+		outer_ip4hdr->fragment_offset = IP_DN_FRAGMENT_FLAG;
+		outer_ip4hdr->time_to_live = IP_DEFTTL;
+		outer_ip4hdr->next_proto_id = IPPROTO_IPIP;
+		/* The source address is the Gatekeeper server IP address. */
+		outer_ip4hdr->src_addr = info->flow.f.v4.src;
+		/* The destination address is the Grantor server IP address. */
+		outer_ip4hdr->dst_addr = info->flow.f.v4.dst;
+
+		outer_ip4hdr->total_length = rte_cpu_to_be_16(pkt->data_len
+				- sizeof(struct ether_hdr));
+
+		/*
+		 * Configure TX offload on a ip-in-ip-encapsulated tcp packet:
+		 * out_eth/out_ip/in_ip/in_tcp/payload
+		 *
+		 * Now, only calculate outer IP checksum. If necessary,
+		 * change here to calculate checksum of out_ip, in_ip, in_tcp.
+		 */
+		outer_ip4hdr->hdr_checksum = 0;
+
+		ol_flags |= (PKT_TX_IPV4 | PKT_TX_IP_CKSUM | PKT_TX_OUTER_IPV4);
+
+		pkt->outer_l2_len = sizeof(struct ether_hdr);
+		pkt->outer_l3_len = sizeof(struct ipv4_hdr);
+		pkt->ol_flags |= ol_flags;
```

Comment:
Do we need a separate `ol_flags` variable, or could we take the bitwise operation above and put it on this line?